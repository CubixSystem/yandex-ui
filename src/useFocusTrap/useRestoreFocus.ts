import { RefObject, useEffect, useRef } from 'react';

import { canUseDOM } from '../lib/canUseDOM';

export interface UseRestoreFocusOptions {
    /**
     * При включении запоминает элемент, на котором находится фокус.
     */
    enabled: boolean | undefined;

    /**
     * Ссылка на элемент, на который нужно вернуть фокус.
     *
     * @default {document.activeElement}
     */
    restoreFocusRef?: RefObject<HTMLElement>;
}

/**
 * Реакт-хук, позволяющий вернуть фокус на последний фокусируемый элемент.
 *
 * @example
 * const Modal: FC<ModalProps> = (props) => {
 *   const { visible, nodeToRestore } = props;
 *
 *   useRestoreFocus({ enabled: visible, nodeToRestore });
 *   ...
 * }
 */
export function useRestoreFocus(options: UseRestoreFocusOptions) {
    const { enabled, restoreFocusRef } = options;
    const initialActiveElement = useRef<HTMLElement | null>(null);
    const ref = useRef<HTMLElement | null>(null);

    if (enabled && !initialActiveElement.current) {
        // React программно вызывает фокус для элементов, у которых есть свойство `autoFocus`
        // поэтому нам важно перехватить активный элемент на этапе рендера
        initialActiveElement.current = canUseDOM() ? (document.activeElement as HTMLElement | null) : null;
    }

    useEffect(() => {
        ref.current = restoreFocusRef ? restoreFocusRef.current : null;
    });

    useEffect(() => {
        if (!enabled) {
            return;
        }

        return () => {
            const element = ref.current || initialActiveElement.current;
            initialActiveElement.current = null;
            ref.current = null;

            if (element && element.focus && document.body.contains(element) && element !== document.activeElement) {
                element.focus();
            }
        };
    }, [enabled]);
}
