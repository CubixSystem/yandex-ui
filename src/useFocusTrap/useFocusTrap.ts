import { RefObject, useEffect, useMemo, useRef } from 'react';

import { createFocusTrap, FocusTrapInstance, FocusTrapOptions } from '../lib/focus-trap';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../useIsomorphicLayoutEffect';
import { useRestoreFocus } from './useRestoreFocus';

export interface UseFocusTrapOptions {
    /**
     * Ссылка на элемент, внутри которого будет действовать ловушка.
     */
    scopeRef: RefObject<HTMLElement>;

    /**
     * Включить ловушку фокуса.
     */
    enabled: boolean | undefined;

    /**
     * Вернуть фокус на триггер после выключения ловушки.
     *
     * @default true
     */
    restoreFocus?: boolean;

    /**
     * Ссылка на элемент, на который нужно вернуть фокус после выключения ловушки.
     */
    restoreFocusRef?: RefObject<HTMLElement>;

    /**
     * Выставить фокус на первом фокусируемом элементе внутри ловушки.
     *
     * @default true
     */
    autoFocus?: boolean;
}

/**
 * Реакт-хук, для создания ловушки фокуса.
 *
 * @example
 * const Modal: FC<ModalProps> = (props) => {
 *   const { visible, children } = props;
 *   const scope = useRef(null);
 *
 *   useFocusTrap({ enabled: visible, scope });
 *
 *   return (
 *     <div ref={scope}>{children}</div>
 *   )
 * }
 */
export function useFocusTrap(options: UseFocusTrapOptions) {
    const { scopeRef, enabled, restoreFocusRef, restoreFocus = true, autoFocus = true } = options;
    const instance = useRef<FocusTrapInstance | null>(null);
    const ref = useRef<HTMLElement | null>(null);

    const trapOptions = useMemo<FocusTrapOptions>(() => {
        return {
            autoFocus,
        };
    }, [autoFocus]);

    useEffect(() => {
        if (instance.current) {
            instance.current.setOptions(trapOptions);
        }
    }, [trapOptions]);

    useLayoutEffect(() => {
        if (ref.current === scopeRef.current) {
            return;
        }

        ref.current = scopeRef.current;

        if (instance.current) {
            instance.current.deactivate();
            instance.current = null;
        }

        if (ref.current) {
            instance.current = createFocusTrap(ref.current, trapOptions);

            if (enabled) {
                instance.current.activate();
            }
        }
    });

    useLayoutEffect(() => {
        if (!enabled || !instance.current) {
            return;
        }

        instance.current.activate();

        return () => {
            if (instance.current) {
                instance.current.deactivate();
            }
        };
    }, [enabled]);

    useRestoreFocus({ enabled: enabled && restoreFocus, restoreFocusRef });
}
