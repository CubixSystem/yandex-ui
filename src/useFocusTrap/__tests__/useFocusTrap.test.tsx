import React, { FC, useRef } from 'react';

import { screen, createClientRender } from '../../internal/testing';
import { useFocusTrap, UseFocusTrapOptions } from '../useFocusTrap';

type FixtureProps = Partial<UseFocusTrapOptions>;

const FocusTrapFixture: FC<FixtureProps> = (props) => {
    const { children, scopeRef, enabled = false, ...options } = props;
    const innerScope = useRef<HTMLDivElement>(null);

    useFocusTrap({ ...options, enabled, scopeRef: scopeRef || innerScope });

    return (
        <div data-testid="wrapper">
            <div ref={innerScope} data-testid="container">
                {children}
            </div>
        </div>
    );
};

describe('useFocusTrap', () => {
    const render = createClientRender();

    test('должен создать элементы-ловушки для контейнера только при включенном состоянии', () => {
        const { setProps } = render<FixtureProps>(<FocusTrapFixture enabled={false} />);

        expect(screen.getByTestId('wrapper')).toMatchSnapshot();

        setProps({ enabled: true });

        expect(screen.getByTestId('wrapper')).toMatchSnapshot();

        setProps({ enabled: false });

        expect(screen.getByTestId('wrapper')).toMatchSnapshot();
    });

    test('должен обновить DOM если поменялась ссылка на контейнер', () => {
        const scopeRef1 = { current: null };
        const scopeRef2 = { current: null };

        const { setProps, unmount } = render<FixtureProps>(
            <FocusTrapFixture enabled scopeRef={scopeRef1}>
                <div ref={scopeRef1} data-testid="container-1" />
                <div ref={scopeRef2} data-testid="container-2" />
            </FocusTrapFixture>,
        );

        expect(screen.getByTestId('wrapper')).toMatchSnapshot();

        setProps({ enabled: true, scopeRef: scopeRef2 });

        expect(screen.getByTestId('wrapper')).toMatchSnapshot();

        setProps({ enabled: true, scopeRef: { current: null } });

        expect(screen.getByTestId('wrapper')).toMatchSnapshot();

        unmount();
    });
});
