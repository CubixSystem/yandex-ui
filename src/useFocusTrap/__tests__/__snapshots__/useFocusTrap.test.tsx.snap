// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFocusTrap должен обновить DOM если поменялась ссылка на контейнер 1`] = `
<div
  data-testid="wrapper"
>
  <div
    data-testid="container"
  >
    <span
      aria-hidden="true"
      data-focus-trap-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="1"
    />
    <div
      data-testid="container-1"
    />
    <span
      aria-hidden="true"
      data-focus-trap-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-testid="container-2"
    />
  </div>
</div>
`;

exports[`useFocusTrap должен обновить DOM если поменялась ссылка на контейнер 2`] = `
<div
  data-testid="wrapper"
>
  <div
    data-testid="container"
  >
    <div
      data-testid="container-1"
    />
    <span
      aria-hidden="true"
      data-focus-trap-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="1"
    />
    <div
      data-testid="container-2"
    />
    <span
      aria-hidden="true"
      data-focus-trap-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`useFocusTrap должен обновить DOM если поменялась ссылка на контейнер 3`] = `
<div
  data-testid="wrapper"
>
  <div
    data-testid="container"
  >
    <div
      data-testid="container-1"
    />
    <div
      data-testid="container-2"
    />
  </div>
</div>
`;

exports[`useFocusTrap должен создать элементы-ловушки для контейнера только при включенном состоянии 1`] = `
<div
  data-testid="wrapper"
>
  <div
    data-testid="container"
  />
</div>
`;

exports[`useFocusTrap должен создать элементы-ловушки для контейнера только при включенном состоянии 2`] = `
<div
  data-testid="wrapper"
>
  <span
    aria-hidden="true"
    data-focus-trap-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="1"
  />
  <div
    data-testid="container"
  />
  <span
    aria-hidden="true"
    data-focus-trap-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
</div>
`;

exports[`useFocusTrap должен создать элементы-ловушки для контейнера только при включенном состоянии 3`] = `
<div
  data-testid="wrapper"
>
  <div
    data-testid="container"
  />
</div>
`;
